pensieve/export_json.py

```diff
@@ -0,0 +1,106 @@
+from google.cloud import bigquery
+from google.cloud import storage
+import logging
+import smart_open
+
+logging.getLogger(__name__)
+
+
+def _get_statistics_tables(client, bq_dataset: str):
+    """Returns statistics table names and their last modified timestamp."""
+    job = client.query(
+        f"""
+        SELECT table_id, TIMESTAMP_MILLIS(last_modified_time) as last_modified
+        FROM {bq_dataset}.__TABLES__
+        WHERE table_id LIKE 'statistics_%'
+    """
+    )
+
+    result = job.result()
+    return {row.table_id: row.last_modified for row in result}
+
+
+def _get_gcs_blobs(storage_client, bucket):
+    """Return all blobs in the GCS location with their last modified timestamp."""
+    blobs = storage_client.list_blobs(bucket)
+
+    return {blob.name.replace(".json", ""): blob.updated for blob in blobs}
+
+
+def _export_table(client, project_id, dataset_id, table, bucket, storage_client):
+    """Export a single table or view to GCS as JSON."""
+    # since views cannot get exported directly, write data into a temporary table
+    job = client.query(
+        f"""
+        SELECT *
+        FROM {dataset_id}.{table}
+    """
+    )
+
+    job.result()
+
+    # get the temporary table results are written to
+    tmp_table = job._properties["configuration"]["query"]["destinationTable"]
+
+    destination_uri = f"gs://{bucket}/{table}.ndjson"
+    dataset_ref = bigquery.DatasetReference(project_id, tmp_table["datasetId"])
+    table_ref = dataset_ref.table(tmp_table["tableId"])
+
+    logging.info(f"Export table {table} to {destination_uri}")
+
+    job_config = bigquery.ExtractJobConfig()
+    job_config.destination_format = "NEWLINE_DELIMITED_JSON"
```

Comment:
BigQuery doesn't support JSON as export format by default. So, the easiest way is to export tables as ndjson and convert it to json.

Reply:
lol, booooooo